### ISSUE TITLE
Join field

### ISSUE THREAD

**Issue by andrei9669** [marmelab/react-admin#6348]
**Is your feature request related to a problem? Please describe.**
You know how we use ReferenceInput to get the references of other resources and such. but what if, we had a join field?
Currently, in GetListParams, there's only pagination, sort and filter, but if we could add a join in there, everything would be much simpler.
my current issue why referenceField isn't enough is, for example, let us say we have 2 tables, sites, and clients. My use-case is that I want to show to the user sites resource but not clients, so in my backend, I just block the clients api point, but in sites, I still want to show the client name, with what the site is related to.

**Describe the solution you'd like**
My desired solution would be that we have a new Field component, called JoinField, or something like that, where you pass the source as in any other component, but in this case, that source field name would be retrievable in dataProvider params and then added to the join in the query.

**Describe alternatives you've considered**
Right now my solution is that I had to practically copy-paste `ra-data-nestjsx-crud` provider and add an if check, for if the resource is "site", then add the join. and then on the List view, I used Function component to get that field.
This example is just one of many things that it could be used for, in this case, I used it to conditionally render link. Yes, I know that ReferenceField has `link` prop which I could set false, but you still get the idea, I hope.
```typescript
// dataProvider.ts
export const dataProvider: DataProvider = {
  ...crudProvider,
  getList: (resource, params) => {
    const { page, perPage } = params.pagination;
    const { q: queryParams, ...filter } = params.filter || {};

   // this part down here <--------------------------------------------------------------
    const join: Array<QueryJoin | QueryJoinArr> = [];
    if (resource === 'site') {
      join.push({ field: 'client' });
    }
   // ----------------------

    const encodedQueryParams = composeQueryParams(queryParams);
    const encodedQueryFilter = RequestQueryBuilder.create({
      filter: composeFilter(filter),
    })
      .setLimit(perPage)
      .setPage(page)
      .sortBy(params.sort as QuerySort)
      .setOffset((page - 1) * perPage)
      .setJoin(join) //  <-------------------------------------------------------------- and this part
      .query();

    const query = mergeEncodedQueries(encodedQueryParams, encodedQueryFilter);

    const url = `${apiUrl}/${resource}?${query}`;

    return httpClient(url).then(({ json }) => ({
      data: json.data,
      total: json.total,
    }));
  },
};
```
```typescript
// SiteList.tsx
const SiteList = ({
  permissions,
  ...props
}: ListProps & { permissions?: Permissions }): React.ReactElement => {
  const canEdit = useCanEdit('site', permissions);
  const canOpenClient = useCanEdit('client', permissions);

  return (
    <List {...props} filters={<SiteFilter />}>
      <Datagrid>
        <TextField source="name" />
        <FunctionField<SiteResource>
          label="Client"
          render={(record) => {
            const linkToUser = linkToRecord(
              '/client',
              record?.client.id,
              'show',
            );
            if (canOpenClient) {
              return (
                <Link to={linkToUser}>
                  <StyledTypography variant="body2">
                    {record?.client.name}
                  </StyledTypography>
                </Link>
              );
            }
            return (
              <StyledTypography variant="body2">
                {record?.client.name}
              </StyledTypography>
            );
          }}
        />
        {canEdit && <EditButton />}
      </Datagrid>
    </List>
  );
};
```

**Additional context**
I think that this solution could actually improve ReferenceField as well since there would be fewer requests made. win-win

There is one more issue with how stuff currently works, maybe I'm just doing it incorrectly. with typeorm, currently, my entity looks like this:
```typescript
@Entity('site', { schema: 'public' })
export class Site {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column('text')
  name: string;

  @Column('uuid')
  clientId: string;

  @ManyToOne(() => Client, (client) => client.sites, {
    onDelete: 'RESTRICT',
    onUpdate: 'CASCADE',
  })
  @JoinColumn([{ name: 'clientId', referencedColumnName: 'id' }])
  client: Client;
}

```
notice the clientId and client. with join, we could just keep the client field.
and in List view I have to use it like this:
```typescript
<ReferenceField reference="client" source="clientId" >
    <TextField source="name" />
</ReferenceField>
```

So not only we could be making fewer requests, but also clean up the backend from extra fields.

**Comment by fzaninotto** [comment]
I'm sorry, I don't understand your problem and use case. You're mentioning "Join", which is a relational database concept. React-admin isn't tied to relational databases. 

So please explain again what you want to do, without assuming we know what a "Join" is.

**Comment by andrei9669** [comment]
I mean it would be nice if we somehow could use this https://github.com/nestjsx/crud/wiki/Requests#join
or how should I put it. hmm, what I want is that instead of using IDs for the reference field, we could send requests to get more data from a single resource.
in the data provider, getList has 2 params, resource and then object with pagination, sort and filter. but I want/need join params or a way to add additional custom params, which would be imo even better.

TL;DR; I want an option/ability to send additional data to my data provider, besides resource, pagination, sort and filter.

**Comment by fzaninotto** [comment]
Thanks for the clarification. 

In my opinion, you're asking your dataProvider to return 2 different sets of data for the same resource. For react-admin, these are 2 different resources. 

So I suggest you add a new `<Resource>` to your admin for the "site with client" resource, and implement the join logic in your dataProvider (as you did). 

As there is already a way to do what you ask without any change in react-admin, we won't implement the ability to pass metadata to queries (which would be a big change).
